user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaProducerService.java

```diff
@@ -92,15 +90,21 @@ public KafkaProducerService(final Sink<ProducerRecord<String, UserRequest>, Comp
                 .collect(Collectors.toList());
 
         // Process requests
-        // Akka reactive streams can't have null elements, so map to and from optional
-        Source.fromJavaStream(() -> requests.stream().map(Optional::ofNullable))
+        return Source
+                // Akka reactive streams can't have null elements, so map to and from optional
+                .fromJavaStream(() -> requests.stream().map(Optional::ofNullable))
+
+                // Create kafka producer record from request
                 .map(request -> new ProducerRecord<String, UserRequest>(topic.getName(), partition, null, request.orElse(null), kafkaHeaders))
+
+                // Sink records to this service's upstream topic (not downstream)
                 .toMat(this.upstreamSink, Keep.right())
+
+                // Run the graph
                 .run(this.materializer)
-                .toCompletableFuture()
-                .join();
 
-        // Return results
-        return new ResponseEntity<>(HttpStatus.ACCEPTED);
+                // Return a CompletableFuture<Void> result
+                .toCompletableFuture()
+                .thenApply(x -> null);
```

Comment:
I assume returning a reasonable status code etc. is still done in the controller?

Reply:
Correct, the controller returns a `202` status code.